File: C++/ЕГЭ2021/25/Решение заданий. Поляков/4215.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
using namespace std;
int f(int x) {
	int mina = 1000000;
	vector<int> div;
	int sqr = round(sqrt(x));
	for (int i = 2; i <= sqr; i++) {
		if (x % i == 0) {
			div.push_back(i);
			div.push_back(x / i);
		}
		for (int j = 0; j < div.size(); j++) {
			if ((div[j] % 10 == 8) and (div[j] != 8) and (div[j] != x)) {
				mina = min(div[j], mina);
			}
		}
	}
	if (mina != 1000000)
		return mina;
	else
		return false;
}


int main() {


	int count = 0;

	for (int i = 500001; i <= 510020; i++) {
		if (count == 5)
			break;
		if (f(i) != false) {
			cout << i << " " << f(i) << endl;
			count++;
		}
	}


	return 0;
}

File: Python/ЕГЭ2021;2022/25/Решение заданий. Поляков/4215.py
def div(n):
    divs = set()

    for d in range(2, int(n**0.5)+1):
        if d == 8:
            continue

        if n % d == 0:
            if d % 10 == 8:
                divs.add(d)

            if (n//d) % 10 == 8:
                divs.add(n//d)

    if len(divs) == 0:
        return 0
    else:
        return min(divs)


c = 0

for i in range(500_001, 10**98):
    r = div(i)

    if r != 0:
        print(i, r)
        c += 1
        
        if c == 5:
            break
